router/_RPC_api.gno

```diff
@@ -47,109 +47,105 @@ func ApiGetRatiosFromBase() string {
+			tokenRatio[token] = avgPriceX96
+
 		}
-		// TOKEN ENDS
 	}
+	// LOOP FIN
 
-	tokenPrices := []TokenPathPrice{}
-	for token, price := range tokenPrice {
-		tokenPrices = append(tokenPrices, TokenPathPrice{
-			Path:  token,
-			Price: price,
+	var tokenRatios = []TokenRatio{}
+	for token, ratio := range tokenRatio {
+		tokenRatios = append(tokenRatios, TokenRatio{
+			Token: token,
+			Ratio: ratio,
 		})
 	}
 
-	return tokenPrices
+	return tokenRatios
 }
 
-func calculateTokenPrice(currentToken, swapPath string, numPools, proceed int, currentPriceX96 bigint) bigint {
-	currentPoolPathKey := makePoolPath(swapPath, proceed)
-	currentPool := p.GetPoolFromPoolPath(currentPoolPathKey)
+func calculateTokenRatio(currentToken string, routes []PoolWithMeta, proceed int, priceX96 bigint) bigint {
+	poolPath := routes[len(routes)-proceed-1].poolPath
+	pool := pl.GetPoolFromPoolPath(poolPath)
 
-	poolToken0 := currentPool.PoolGetToken0Path()
-	poolToken1 := currentPool.PoolGetToken1Path()
+	token0Path := pool.PoolGetToken0Path()
+	token1Path := pool.PoolGetToken1Path()
 
-	if poolToken0 == currentToken {
-		currentSqrtPriceX96 := currentPool.PoolGetSlot0SqrtPriceX96()
+	if token1Path == currentToken {
+		poolSqrtPriceX96 := pool.PoolGetSlot0SqrtPriceX96()
 
-		currentPriceX96 *= currentSqrtPriceX96 * currentSqrtPriceX96
-		currentToken = poolToken1
-	} else if poolToken1 == currentToken {
-		currentTick := currentPool.PoolGetSlot0Tick()
-		oppositeTick := -currentTick
+		priceX96 *= poolSqrtPriceX96 * poolSqrtPriceX96
+		currentToken = token0Path
+	} else if token0Path == currentToken {
+		poolTick := pool.PoolGetSlot0Tick()
+		oppositeTick := -poolTick
 		oppositeSqrtPriceX96 := common.TickMathGetSqrtRatioAtTick(oppositeTick)
 
-		currentPriceX96 *= oppositeSqrtPriceX96 * oppositeSqrtPriceX96
-		currentToken = poolToken0
+		priceX96 *= oppositeSqrtPriceX96 * oppositeSqrtPriceX96
+		currentToken = token1Path
 	} else {
-		panic("[ROUTER] _RPC_api.gno__calculateTokenPrice() || wrong condition")
+		// wrong condition
+		// panic("[ROUTER] _RPC_api.gno__calculateTokenRatio() || wrong condition")
+		return 0
+
 	}
 
-	if proceed == numPools-1 {
-		for {
-			if currentPriceX96 < (consts.Q96 * 2) {
-				return currentPriceX96
+	proceed += 1
+
+	if proceed == len(routes) { // numHops
+		for { // remove as much X96 as possible
+			tempPriceX96 := priceX96
+			priceX96 /= consts.Q96
+
+			if priceX96 < consts.MIN_PRICE {
+				return tempPriceX96
 			}
-			currentPriceX96 /= consts.Q96
 		}
 	}
 
-	return calculateTokenPrice(currentToken, swapPath, numPools, proceed+1, currentPriceX96)
-}
-
-func sqrt(x bigint, n int) bigint {
-	result := bigint(1)
-	for i := 0; i < n; i++ {
-		result *= x
-	}
-	return result
+	return calculateTokenRatio(currentToken, routes, proceed, priceX96)
 }
 
 func getTokenList() []string {
 	seen := make(map[string]bool)
 	uniqueTokenList := []string{}
-	poolList := p.PoolGetPoolList()
+	poolList := pl.PoolGetPoolList()
```

Comment:
We don't know the final size of the `uniqueTokenList`, but using the size of the `poolList` seems to allow us to set the initial capacity and this will reduce the re-allocations.

```diff
+ poolList := pl.PoolGetPoolList()
+ uniqueTokenList := make([]string, 0, len(poolList)*2)
- uniqueTokenList := []string{}
- poolList := pl.PoolGetPoolList()
```

Reply:
looks great, but why does initial length requires `* 2`?

Reply:
> looks great, but why does initial length requires `* 2`?

I simply set it to double because I assumed that since each pool has two tokens (`token0`, `token1`), each pool would add up to two unique tokens to the list. However, since this value was not set after exact measurement, we will need to reevaluate this part later.